gotsysctl/gotsysctl.8
.\"
.\" Copyright (c) 2025 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTSYSCTL 8
.Os
.Sh NAME
.Nm gotsysctl
.Nd control the Game of Trees System Daemon
.Sh SYNOPSIS
.Nm
.Op Fl hV
.Op Fl f Ar path
.Ar command
.Op Ar arg ...
.Sh DESCRIPTION
.Nm
controls the
.Xr gotsysd 8
daemon.
.Pp
.Nm
connects to the
.Xr gotsysd 8
unix socket, sends a command, and displays the response from
.Xr gotsysd 8 .
Access to the
.Xr gotsysd 8
unix socket is restricted to specific user accounts,
see
.Xr gotsysd.conf 5 .
.Pp
The options for
.Nm
are as follows:
.Bl -tag -width Ds
.It Fl h
Display usage information and exit immediately.
.It Fl f Ar path
Set the
.Ar path
to the unix socket which
.Xr gotsysd 8
is listening on.
If not specified, the path
.Pa /var/run/gotsysd.sock
will be used.
.It Fl V , -version
Display program version and exit immediately.
.El
.Pp
The commands for
.Nm
are as follows:
.Bl -tag -width Ds
.It Cm info
Display information about a running
.Xr gotsysd 8
instance.
.El
.Sh SEE ALSO
.Xr gotsys.conf 5 ,
.Xr gotsysd.conf 5 ,
.Xr gotsysd 8
.Sh AUTHORS
.An Stefan Sperling Aq Mt stsp@openbsd.org


CHANGES
* got 0.110; 2025-02-20
- fix an endless loop in got-read-pack (regression from 0.109)
- change gotwebd diff algorithm from Myers to Patience diff
- gotd regress depends on p5-Net-Daemon now due to an added test case

* got 0.109; 2025-02-14
- fix gotd failing to protect references when the client sends an empty pack
- during pack generation, fix exclusion of commits via an ancestor commit
- fix a bogus "received unexpected privsep message" error from gotsh
- fix diffstat path order bug in field width computation
- gotwebd: preserve 'folder=' parameter when following More links

* got 0.108; 2025-01-22
- add ssh -i identity-file support to commands which use the network
- make 'got import' output independent of readdir(3) entry order
- avoid full file content comparisons in 'got status' for speed
- tog: fix NULL deref when log view T keymap is used on worktree entry
- tog: fix a deadlock (hang) in the log view implementation
- tog: plug a memory leak
- tog: do not exit if a tag pointing at a non-commit is selected in ref view
- tog: do not mark an incorrect base commit in nested log views
- tog: fix NULL deref when scrolling small tree views down
- tog: avoid showing a negative log view entry index
- tog: do not apply a pointless count modifier to the H, &, p keymaps
- tog: do not make users wait for the worktree diff to quit out of tog
- gotwebd: make parent process drop root privileges
- gotwebd: drop read access to /var/www from parent process
- gotwebd: rename "socket" processes to "server"
- gotadmin cleanup: pack the repository before removing objects
- gotadmin cleanup: do not delete directly referenced trees and blobs
- gotadmin cleanup: do not delete objects reachable via nested tags
- regress: skip test memleak_send_basic in sha256 mode; expected to fail
- regress: make seq(1) invocations portable to fix test failures on linux
- regress/gotwebd: implement paginated commits test

* got 0.107; 2024-12-28
- gotwebd.css styling tweaks
- hide ssh debug output during fetch/send -v, keep showing it at -vv and -vvv
- discern mixed-commit worktree diffs with commit ID headers
- gotwebd: avoid printf("%s", NULL) when path parameter is not in query
- implement a regression test harness for gotwebd
- fix free() called with bogus pointer in 'got fetch'; regression from 0.106
- ensure config privsep children get collected upon error to prevent zombies
- fix some fprintf(3) failure checks
- gotwebd: replace strftime(3) with asctime_r(3) for the sake of consistency
- tweak gotwebd log message levels, and log requests in verbose (-v) mode
- prevent out-of-bounds read during gotwebd fcgi record debugging
- implement tog work tree diff support via log view and CLI
- improve error reporting when 'got patch' encounters malformed patches
- improve got_opentemp_named_fd error reporting by showing the path template
- add ssh -J jumphost support to got and cvg commands which use the network
- add regression tests checking for memory leaks with Otto malloc and ktrace
- got tag: change -s signer to -S signer
- got tag: provide one-line output mode via new -s option
- tog: use wtimeout(3) instead of nodelay(3) to honour our display refresh rate 
- switch got_pathlist data store from TAILQ to RB-tree
- plug many memory leaks, some of which affected gotwebd in particular

* got 0.106; 2024-11-21
- prevent gotd from exiting with pending notifications if client disconnects
- convert got to the new imsg API
- gotwebd: improve performance of repository age calculations
- gotwebd: ensure child processes inherit non-default config

* got 0.105; 2024-11-14
- fix bogus "branch on server has different ancestry" errors from 'got send'
- do not try to merge binary files during (un)stage -p and revert -p
- fix gotd notifications about changes involving empty files
- minor gotwebd.css styling changes
- plug several memory and file-descriptor leaks

* got 0.104; 2024-10-22
  see git repository history for per-change authorship information
- gotd.conf: document the macro syntax
- tog: prevent a segfault upon unexpected object type in ref list view
- fix pack file creation in the presence of tagged tag objects
- plugged some memory leaks
- fix a crash when unstaging a file which has been removed from disk
- gotwebd: fix out of bounds access while handling the configuration

* got 0.103; 2024-09-24
  see git repository history for per-change authorship information
- fix bug causing performance to degrade as more and more pack files appear
- tog: add diff view 'p' keymap to write the diff to file
- tog: display diffstat in diff view when diffing blobs or trees directly
- gotwebd: show commit id prefix on briefs page
- add support for HMAC digests to gotd HTTP notifications
- move authentication credentials from gotd.conf(5) to gotd-secrets.conf(5)
- fix spurious tog regression test failures on slower machines
- restore abort() calls in lib/hash.c to quiet potential compiler warnings
- gotwebd: unbreak listing of tags on the summary page (regression from 0.102)
- gotwebd: minor tweaks to the HTML for ease of styling

* got 0.102; 2024-08-14
  see git repository history for per-change authorship information
- support for sha256 repositories; the network protocol requires git(1) for now
- gotwebd: add support for the "owner" file
- gotwebd: fix the README link in the summary view
- fix handling of .gitignore files containing empty lines
- fix handling of files without trailing newline in histedit, rebase and merge
- gotd: allow numeric UIDs in permit/deny rules, as intended
- gotd: support numeric UIDs in the `user' directive
- fix comment handling and explain quoting in the *.conf.5 man pages
- tog: add ability to mark arbitrary commits to diff them
- print file index and work tree version in got info

* got 0.101; 2024-07-11
  see git repository history for per-change authorship information
- improve gotwebd.8 one-line description
- adjust some SIZE_MAX bounds checks

gotsysctl/gotsysctl.c
/*
 * Copyright (c) 2025 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/tree.h>
#include <sys/signal.h>
#include <sys/socket.h>
#include <sys/un.h>

#include <err.h>
#include <event.h>
#include <imsg.h>
#include <limits.h>
#include <locale.h>
#include <sha1.h>
#include <sha2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <vis.h>

#include "got_error.h"
#include "got_version.h"
#include "got_path.h"
#include "got_opentemp.h"
#include "got_repository.h"
#include "got_reference.h"
#include "got_object.h"

#include "got_lib_poll.h"

#include "gotsysd.h"

#ifndef nitems
#define nitems(_a)	(sizeof((_a)) / sizeof((_a)[0]))
#endif

static volatile sig_atomic_t sigint_received;

static void
catch_sigint(int signo)
{
	sigint_received = 1;
}

struct gotsysctl_cmd {
	const char	*cmd_name;
	const struct got_error *(*cmd_main)(int, char *[], int);
	void		(*cmd_usage)(void);
};

__dead static void	usage(int, int);

__dead static void	usage_info(void);

static const struct got_error*		cmd_info(int, char *[], int);

static const struct gotsysctl_cmd gotsysctl_commands[] = {
	{ "info",	cmd_info,	usage_info },
};

__dead static void
usage_info(void)
{
	fprintf(stderr, "usage: %s info\n", getprogname());
	exit(1);
}

static const struct got_error *
show_info(struct imsg *imsg)
{
	const struct got_error *err = NULL;
	struct gotsysd_imsg_info info;
	size_t datalen;
	char *repo_dir_safe = NULL, *commit_id_str = NULL;

	datalen = imsg->hdr.len - IMSG_HEADER_SIZE;
	if (datalen != sizeof(info))
		return got_error(GOT_ERR_PRIVSEP_LEN);
	memcpy(&info, imsg->data, sizeof(info));

	info.repository_directory[sizeof(info.repository_directory) - 1] = '\0';

	if (info.commit_id.algo == GOT_HASH_SHA1 ||
	    info.commit_id.algo == GOT_HASH_SHA256) {
		err = got_object_id_str(&commit_id_str, &info.commit_id);
		if (err)
			return err;
	}

	if (stravis(&repo_dir_safe, info.repository_directory,
	    VIS_SAFE) == -1)
		return got_error_from_errno("stravis");

	printf("gotsysd PID: %d\n", info.pid);
	printf("verbosity: %d\n", info.verbosity);
	printf("repository directory: %s\n", repo_dir_safe);
	printf("UID range: %u-%u\n", info.uid_start, info.uid_end);
	printf("gotsys.conf commit: %s\n",
	    commit_id_str ? commit_id_str : "unknown");

	free(repo_dir_safe);
	free(commit_id_str);
	return NULL;
}

static const struct got_error *
apply_unveil_none(void)
{
	if (unveil("/", "") == -1)
		return got_error_from_errno("unveil");

	if (unveil(NULL, NULL) == -1)
		return got_error_from_errno("unveil");

	return NULL;
}

static const struct got_error *
cmd_info(int argc, char *argv[], int gotsysd_sock)
{
	const struct got_error *err;
	struct imsgbuf ibuf;
	struct imsg imsg;

	err = apply_unveil_none();
	if (err)
		return err;
#ifndef PROFILE
	if (pledge("stdio", NULL) == -1)
		return got_error_from_errno("pledge");
#endif
	if (imsgbuf_init(&ibuf, gotsysd_sock) == -1)
		return got_error_from_errno("imsgbuf_init");

	if (imsg_compose(&ibuf, GOTSYSD_IMSG_CMD_INFO, 0, 0, -1,
	    NULL, 0) == -1) {
		err = got_error_from_errno("imsg_compose INFO");
		goto done;
	}

	err = gotsysd_imsg_flush(&ibuf);
	if (err)
		goto done;

	err = gotsysd_imsg_poll_recv(&imsg, &ibuf, 0);
	if (err)
		goto done;

	switch (imsg.hdr.type) {
	case GOTSYSD_IMSG_ERROR:
		err = gotsysd_imsg_recv_error(NULL, &imsg);
		break;
	case GOTSYSD_IMSG_INFO:
		err = show_info(&imsg);
		break;
	default:
		err = got_error(GOT_ERR_PRIVSEP_MSG);
		break;
	}

	imsg_free(&imsg);
done:
	imsgbuf_clear(&ibuf);
	return err;
}

static void
list_commands(FILE *fp)
{
	size_t i;

	fprintf(fp, "commands:");
	for (i = 0; i < nitems(gotsysctl_commands); i++) {
		const struct gotsysctl_cmd *cmd = &gotsysctl_commands[i];
		fprintf(fp, " %s", cmd->cmd_name);
	}
	fputc('\n', fp);
}

__dead static void
usage(int hflag, int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;

	fprintf(fp, "usage: %s [-hV] [-f path] command [arg ...]\n",
	    getprogname());
	if (hflag)
		list_commands(fp);
	exit(status);
}

static int
connect_gotsysd(const char *socket_path)
{
	int gotsysd_sock = -1;
	struct sockaddr_un sun;

	if ((gotsysd_sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		err(1, "socket");

	memset(&sun, 0, sizeof(sun));
	sun.sun_family = AF_UNIX;
	if (strlcpy(sun.sun_path, socket_path, sizeof(sun.sun_path)) >=
	    sizeof(sun.sun_path))
		errx(1, "gotsysd socket path too long");
	if (connect(gotsysd_sock, (struct sockaddr *)&sun, sizeof(sun)) == -1)
		err(1, "connect: %s", socket_path);

	return gotsysd_sock;
}

int
main(int argc, char *argv[])
{
	const struct gotsysctl_cmd *cmd;
	int gotsysd_sock = -1, i;
	int ch;
	int hflag = 0, Vflag = 0;
	static const struct option longopts[] = {
	    { "version", no_argument, NULL, 'V' },
	    { NULL, 0, NULL, 0 }
	};
	const char *socket_path = GOTSYSD_UNIX_SOCKET;

	setlocale(LC_CTYPE, "");

	signal(SIGINT, catch_sigint);

#ifndef PROFILE
	if (pledge("stdio rpath wpath cpath unix unveil", NULL) == -1)
		err(1, "pledge");
#endif

	while ((ch = getopt_long(argc, argv, "+hf:V", longopts, NULL)) != -1) {
		switch (ch) {
		case 'h':
			hflag = 1;
			break;
		case 'f':
			socket_path = optarg;
			break;
		case 'V':
			Vflag = 1;
			break;
		default:
			usage(hflag, 1);
			/* NOTREACHED */
		}
	}

	argc -= optind;
	argv += optind;
	optind = 1;
	optreset = 1;

	if (Vflag) {
		got_version_print_str();
		return 0;
	}

	if (argc <= 0)
		usage(hflag, hflag ? 0 : 1);

	for (i = 0; i < nitems(gotsysctl_commands); i++) {
		const struct got_error *error;

		cmd = &gotsysctl_commands[i];

		if (strncmp(cmd->cmd_name, argv[0], strlen(argv[0])) != 0)
			continue;

		if (hflag)
			cmd->cmd_usage();

		gotsysd_sock = connect_gotsysd(socket_path);
		if (gotsysd_sock == -1)
			return 1;
		error = cmd->cmd_main(argc, argv, gotsysd_sock);
		close(gotsysd_sock);
		if (error) {
			fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
			return 1;
		}

		return 0;
	}

	fprintf(stderr, "%s: unknown command '%s'\n", getprogname(), argv[0]);
	list_commands(stderr);
	return 1;
}
